High-Scale IPcache: Chapter 3

pchaigno opened the pull request:
This is in the context of the high-scale ipcache feature described at https://github.com/cilium/design-cfps/pull/7.

This pull request extends the datapath to be able to encapsulate and decapsulate the special VXLAN tunnel of high-scale ipcache. In this special tunnel, the outer IP addresses are the same as the inner IP addresses; the encapsulation header is just used to carry the source security identity.

This encapsulation therefore allows us to enforce ingress network policies in high-scale ipcache mode. The end-to-end test will however be updated in a separate pull request to reflect that because we still need https://github.com/cilium/cilium/pull/25436 to populate the World CIDR map (see first commit).

Updates: https://github.com/cilium/cilium/issues/25243.
```
Support ingress network policies in high-scale ipcache mode.
```

pchaigno committed:
bpf, maps: Create world CIDR map

This commit adds a new LPM map for world CIDRs. Only IPv4 is supported
for now. The map will be populated via a new CRD in a separate patchset.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
loader, daemon: Create tunneling device for high-scale ipcache

Similarly to the egress gateway feature, the high-scale ipcache mode
requires the tunneling device even if we're running with native routing.
This is because we use that device to send pod-to-pod traffic
encapsulated with the pod IPs (just to carry the identity).

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
bpf: Cover ENABLE_HIGH_SCALE_IPCACHE in compile tests

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
bpf, option: Encapsulate traffic with pod IPs

In case the high-scale ipcache mode is enabled, we want to encapsulate
pod-to-pod traffic with the pod IP address. The goal of this
encapsulation is simply to carry the security identity for the source.

Previous commits introduced a new CRD and its corresponding map. They
tell us which IP addresses belong to entities outside of the cluster, or
at least entities to which we shouldn't encapsulate traffic when in
high-scale IPcache mode.

This commit makes use of the new map. We perform an LPM lookup into the
map to know if we should encapsulate traffic or not.

We also need to take this into account when computing the MTU. Thus, the
TunnelExists() function is updated to return true when the high-scale
ipcache mode is enabled.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>
Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

pchaigno committed:
bpf: Decapsulate traffic encapsulated with pod IPs

When the high-scale ipcache is enabled, we will receive traffic
encapsulated with the pod IP addresses on the native device. Since the
destination IP is not assigned to the host (but to a container), the
Linux stack won't demultiplex it to the overlay device (e.g.,
cilium_vxlan). Instead, the packets will follow their normal way to the
container, via cilium_host if endpoint routes are disabled.

We want to decapsulate the packet before they reach the lxc devices.
We could decapsulate in bpf_lxc, but then the packet paths would be
assymetric. This commit adds support for decapsulation in cilium_host.
Note that will only work when endpoint routes are disabled.

Therefore, in bpf_host, we filter all incoming VXLAN traffic based on
the UDP port and remove the first IP, UDP, VXLAN, and Ethernet headers.
We also parse the source security identity from the VXLAN header. We
then redirect the packet to its expected path, via cilium_vxlan, with
the security identity in skb->cb.

In bpf_overlay, we need a special case for the high-scale ipcache, to
retrieve the security identity from skb->cb instead of getting it from
the tunnel metadata as usual (since packet is already decapsulated at
this point).

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
probe, daemon: Probe for the ability to set outer source IP

This commit defines a new kernel probe to check that the
bpf_skb_set_tunnel_key BPF helper can be used to set the outer source IP
address. This new probe is then used to fatal if it isn't supported and
high-scale IPcache mode is enabled. IPcache mode will require this
kernel feature (see subsequent commit).

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
bpf: Set outer source IP to pod IP

We can use the recent extension of the bpf_skb_set_tunnel_key BPF helper
to set the outer source IP of encapsulated packets. When the high-scale
IPcache mode is enabled, we want to set that outer source IP to the
source pod IP address.

We don't set the outer source IP from XDP as this is only relevant for
traffic from pods.

For older kernels, we need to pass a smaller bpf_tunnel_key struct to
the helper because they don't support the larger struct with the source
IP.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>
Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1324,6 +1326,18 @@ int cil_from_netdev(struct __ctx_buff *ctx)
 int cil_from_host(struct __ctx_buff *ctx)
 {
+#ifdef ENABLE_HIGH_SCALE_IPCACHE
+	int ret = CTX_ACT_OK;
```
Is this needed for a later patch? Right now it's a bit confusing (given we never use the value)

julianwiedmann commented on bpf/bpf_overlay.c:
```diff
@@ -76,7 +77,11 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
+		 * don't need it since we only want to retrieve the tunnel ID anyway.
+		 */
+		key_size = offsetof(struct bpf_tunnel_key, local_ipv4);
+		if (unlikely(ctx_get_tunnel_key(ctx, &key, key_size, 0) < 0))
```
Can we introduce some macro `TUNNEL_KEY_WITHOUT_SRC_IP` for this, and add the description there? Right now we're duplicating it across every user ...

julianwiedmann commented on bpf/include/linux/vxlan.h:
```diff
@@ -0,0 +1,20 @@
+/* SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note */
+/* Copyright Authors of the Linux kernel */
```
I believe this doesn't apply to non-`uapi` kernel headers, so we can't just copy the whole file.

Why not add the actual struct in `common.h` for now, as we recently did for `struct genevehdr` ? (FWIW, that's what I planned for https://github.com/cilium/cilium/pull/24422/commits/0e3952c88fb37920a8cd788ed102d0a1794629db#diff-78122dbc1bf6a53a08f02c035e071f869eaa123b52822d789b79c04ff1cbdf69).

julianwiedmann commented on bpf/lib/overloadable_skb.h:
```diff
@@ -206,10 +208,20 @@ ctx_set_encap_info(struct __sk_buff *ctx, __u32 node_id, __u32 seclabel,
+		 * thus expect a smaller structure.
+		 */
+		key_size = offsetof(struct bpf_tunnel_key, local_ipv4);
+	}
```
For readability, can we initialize the `key_size` with the smaller/old value and then make this

```suggestion
	if (src_ip != 0) {
		key.local_ipv4 = bpf_ntohl(src_ip);
		key_size = sizeof(key);
	}
```

julianwiedmann reviewed (commented):
One licensing aspect, and some (non-blocking) cleanups.

asauber reviewed (approved):
Config-related changes LGTM

julianwiedmann reviewed (approved):
Looks all good, thank you!

rgo3 commented on pkg/probe/probe.go:
```diff
@@ -80,3 +83,58 @@ func HaveIPv6Support() bool {
+// fields will simply not be used. Instead, we set the outer source IP and
+// retrieve it with bpf_skb_get_tunnel_key right after. If the retrieved value
+// equals the value set, we have a confirmation the kernel supports it.
+func HaveOuterSourceIPSupport() (bool, error) {
```
I just noticed we have two packages for feature probing. Could we add this function to `pkg/datapath/linux/probes` instead to keep adding feature probes to only one of them? I'd offer to take care of migrating the rest of the probes existing in `pkg/probe`.

With regards to the API, I would like to suggest to only return an error like all the other feature probing APIs in `pkg/datapath/linux/probes` with following return semantics:
* err == nil: feature is supported
* errors.Is(err, probes.ErrNotSupported): feature is not supported

If we encounter unexpected errors while executing the feature probe itself we fatal in the probe function itself. This would slightly simplify the calling code as well:
```golang
if err := HaveOuterSourceIPSupport(); err != nil {
    log.Fatal(...)
}
```

Not sure if we'd ever have to call this feature probe in a different place than this, but if we do it would be nice to cache the returned error with a global var and only run the actual probe once with `sync.Once`. 

Lastly, could we add a small test for this that checks if it succeeds on kernels where this is supported? We just landed `testutils.SkipOnOldKernel()` for tests like that.

pchaigno replied:
> If we encounter unexpected errors while executing the feature probe itself we fatal in the probe function itself. This would slightly simplify the calling code as well:

I'm not exactly sure what you meant by that. That's already the case. We fatal in the caller if it fails. We have two cases because I want to specialize the message depending on whether the feature is missing vs. probing failed.

> Not sure if we'd ever have to call this feature probe in a different place than this, but if we do it would be nice to cache the returned error with a global var and only run the actual probe once with sync.Once.

I'd be surprised if we ever need this probe somewhere else. Setting our own source IP for the tunnel is very specific to the high-scale IPcache feature. Also, a bit fearful that if I do that, the next reviewer is then going to ask me to avoid global variables :laughing: So skipped this for now.

<br>

I moved the probe, reworked its return semantics, and added a test for it.

rgo3 replied:
> That's already the case. We fatal in the caller if it fails. We have two cases because I want to specialize the message depending on whether the feature is missing vs. probing failed.

I was trying to say that I would move the fataling in case of probe failures into the probe directly to stay consistent across all probing APIs. So like [HaveProgramType](https://github.com/cilium/cilium/blob/3e805306fc4125fca3398397acdf11a13f1a9600/pkg/datapath/linux/probes/probes.go#L358), if an unexpected error occurs, we call log.Fatal directly in the probe.

Everything else LGTM! :rocket:

pchaigno committed:
test: Populate world CIDR map in high-scale ipcache test

Now that the datapath encapsulation and decapsulation are implemented,
the end-to-end test will fail. To decide whether we should encapsulate
or not, we lookup the world CIDR map. If no match is found, we
encapsulate. That means that we will currently always encapsulate (even
when e.g. resolving a domain name from 8.8.8.8) because the world CIDR
map is not currently populated.

The world CIDR map will be populate once the patchset introducing the
new CRD is merged. Until then, we can add a catch-all 0.0.0.0/0 entry to
not encapsulate anything. This commit can be reverted once the
CiliumWorldCIDRSet CRD is merged.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

viktor-kurchenko commented on pkg/probe/probe.go:
```diff
@@ -80,3 +83,58 @@ func HaveIPv6Support() bool {
 	return true
 }
+
+// HaveSourceOuterIPSupport tests whether the kernel support setting the outer
```
Typo: `HaveSourceOuterIPSupport` -> `HaveOuterSourceIPSupport` ?

viktor-kurchenko commented on pkg/maps/worldcidrsmap/worldcidrs.go:
```diff
@@ -0,0 +1,140 @@
+	return initWorldCIDRsMap(MapName4, false)
+}
+
+// initworldCIDRsMap initializes the world CIDR map.
```
Typo: `initworldCIDRsMap` -> `initWorldCIDRsMap`

pchaigno replied:
How did you even spot that? :exploding_head:

pippolo84 reviewed (approved):
LGTM for agent-related changes.
One question: have you considered adding the new `worldcidrsmap` to the `maps` cell in `pkg/maps`? This could ease the config options management and the testing. Might be something to consider for a follow-up PR.
